include/deal.II/base/symmetric_tensor.h

```diff
@@ -2552,6 +2557,117 @@ identity_tensor ()
 
 
 /**
+ * Invert a symmetric rank-2 tensor.
+ *
+ * @note If a tensor is not invertible, then the result is unspecified, but will
+ * likely contain the results of a division by zero or a very small number at
+ * the very least.
+ *
+ * @relates SymmetricTensor
+ * @author Jean-Paul Pelteret, 2016
+ */
+template <int dim, typename Number>
+inline
+SymmetricTensor<2,dim,Number>
+invert (const SymmetricTensor<2,dim,Number> &t)
+{
+  // if desired, take over the
+  // inversion of a 4x4 tensor
+  // from the FullMatrix
+  AssertThrow (false, ExcNotImplemented());
+
+  return SymmetricTensor<2,dim,Number>();
+}
+
+
+
+#ifndef DOXYGEN
+
+template <typename Number>
+inline
+SymmetricTensor<2,1,Number>
+invert (const SymmetricTensor<2,1,Number> &t)
+{
+  SymmetricTensor<2,1,Number> tmp;
+
+  tmp[0][0] = 1.0/t[0][0];
+
+  return tmp;
+}
+
+
+
+template <typename Number>
+inline
+SymmetricTensor<2,2,Number>
+invert (const SymmetricTensor<2,2,Number> &t)
+{
+  SymmetricTensor<2,2,Number> tmp;
+
+  // Sympy result: ([
+  // [ t11/(t00*t11 - t01**2), -t01/(t00*t11 - t01**2)],
+  // [-t01/(t00*t11 - t01**2),  t00/(t00*t11 - t01**2)]  ])
+  const TableIndices<2> idx_00 (0,0);
+  const TableIndices<2> idx_01 (0,1);
+  const TableIndices<2> idx_11 (1,1);
+  const Number det_t
+    = t[idx_00]*t[idx_11]
+      - t[idx_01]*t[idx_01];
+  const double inv_det_t = 1.0/det_t;
+  tmp[idx_00] = t[idx_11]*inv_det_t;
+  tmp[idx_01] = -t[idx_01]*inv_det_t;
+  tmp[idx_11] = t[idx_00]*inv_det_t;
```

Comment:
Would you mind writing this as first the assignment of the elements, and at the very end the multiplication by `inv_det_t` via `operator *=`? This way, you may expose more opportunities for vectorization.

Same for the 3d case below.